File: sub-projects/intelitor-v5.2/lib/cepaf/src/Cepaf/Mesh/OodaSupervisor.fs
// =============================================================================
// OodaSupervisor.fs - SIL-4 Compliant OODA Biomorphic Supervisor
// =============================================================================
// STAMP: SC-SIL4-007, SC-BIO-001, SC-OODA-001, SC-OODA-005, SC-CTRL-006
// AOR: AOR-SIL4-004, AOR-BIO-001, AOR-CAE-001, AOR-COG-001
//
// ## Techniques Implemented
// | Technique | Source | Purpose |
// |-----------|--------|---------|
// | OODA Loop | John Boyd | Decision-action cycle |
// | Hysteresis | Control Theory | Prevent oscillation |
// | Biomorphic Scaling | Metabolic Theory | Resource adaptation |
// | Guardian Validation | Indrajaal Constitution | Safety enforcement |
//
// ## Document Control
// | Field | Value |
// |-------|-------|
// | Version | 1.0.0 |
// | Created | 2026-01-04 |
// | Author | Cybernetic Architect |
// =============================================================================

namespace Cepaf.Mesh

open System
open System.Diagnostics
open System.Threading
open System.Collections.Generic

/// <summary>
/// OODA phase enumeration
/// </summary>
type OodaPhase =
    | Observe
    | Orient
    | Decide
    | Act
    | Verify

/// <summary>
/// Observation data from sensors
/// </summary>
type Observation = {
    /// Timestamp of observation
    Timestamp: DateTimeOffset
    /// Container health states
    ContainerHealth: Map<string, ContainerHealth>
    /// Port availability
    PortsAvailable: bool
    /// Database connectivity
    DatabaseConnected: bool
    /// Network latency (ms)
    NetworkLatencyMs: int
    /// Memory pressure (0.0 - 1.0)
    MemoryPressure: float
    /// CPU utilization (0.0 - 1.0)
    CpuUtilization: float
    /// Active connections per container
    ActiveConnections: Map<string, int>
    /// Error count in last cycle
    RecentErrors: int
    /// SLA compliance status
    SlaCompliant: bool
}

/// <summary>
/// Orientation (analysis) result
/// </summary>
type Orientation = {
    /// Current system health score (0.0 - 1.0)
    HealthScore: float
    /// Trend direction (-1 degrading, 0 stable, +1 improving)
    Trend: int
    /// Identified issues
    Issues: string list
    /// Recommended priorities
    Priorities: string list
    /// Risk level (0 = none, 1 = low, 2 = medium, 3 = high, 4 = critical)
    RiskLevel: int
    /// Degraded containers
    DegradedContainers: string list
    /// 5-order effect prediction
    PredictedEffects: Map<int, string list>
}

/// <summary>
/// Decision (action plan)
/// </summary>
type Decision =
    | NoAction of reason: string
    | BootMesh of config: BootConfig
    | ShutdownMesh of config: ShutdownConfig
    | RestartContainer of id: string * reason: string
    | ScaleUp of count: int
    | ScaleDown of count: int
    | EmergencyStop of reason: string
    | HealthCheck of ids: string list
    | DrainContainer of id: string

/// <summary>
/// Action result
/// </summary>
type ActionResult =
    | Success of message: string * durationMs: int64
    | Failure of error: string * durationMs: int64
    | Skipped of reason: string
    | Vetoed of guardianReason: string

/// <summary>
/// OODA cycle result
/// </summary>
type OodaCycle = {
    /// Cycle number
    CycleNumber: int
    /// Start time
    StartTime: DateTimeOffset
    /// End time
    EndTime: DateTimeOffset
    /// Total duration in milliseconds
    DurationMs: int64
    /// Observation collected
    Observation: Observation
    /// Orientation result
    Orientation: Orientation
    /// Decision made
    Decision: Decision
    /// Action result
    ActionResult: ActionResult
    /// SLA compliant (< 100ms per SC-BIO-001)
    SlaCycleCompliant: bool
}

/// <summary>
/// Supervisor configuration
/// </summary>
type SupervisorConfig = {
    /// OODA cycle interval in milliseconds
    CycleIntervalMs: int
    /// Maximum cycle duration (SLA) in milliseconds
    MaxCycleDurationMs: int
    /// Hysteresis margin for decisions (prevents oscillation)
    HysteresisMargin: float
    /// Hold cycles before reversing decision
    HysteresisHoldCycles: int
    /// Enable Guardian validation
    EnableGuardian: bool
    /// Health score threshold for action
    HealthThreshold: float
    /// Maximum consecutive failures before emergency
    MaxConsecutiveFailures: int
    /// Enable verbose logging
    Verbose: bool
}

/// <summary>
/// Supervisor state
/// </summary>
type SupervisorState = {
    /// Digital twin reference
    Twin: DigitalTwin
    /// Configuration
    Config: SupervisorConfig
    /// Cycle history (last N cycles)
    History: OodaCycle list
    /// Current phase
    CurrentPhase: OodaPhase
    /// Consecutive failure count
    ConsecutiveFailures: int
    /// Last decision (for hysteresis)
    LastDecision: Decision option
    /// Cycles since last decision change
    CyclesSinceDecisionChange: int
    /// Running flag
    Running: bool
}

/// <summary>
/// OODA Supervisor operations module
/// </summary>
module OodaSupervisor =

    /// Default supervisor configuration (SC-BIO-001: <100ms cycles)
    let defaultConfig : SupervisorConfig = {
        CycleIntervalMs = 30000      // 30s between cycles (per AOR-BIO-001)
        MaxCycleDurationMs = 100     // 100ms max per SC-OODA-001
        HysteresisMargin = 0.1       // 10% margin per SC-OODA-005
        HysteresisHoldCycles = 3     // Hold 3 cycles per SC-OODA-005
        EnableGuardian = true
        HealthThreshold = 0.8        // 80% health required
        MaxConsecutiveFailures = 3
        Verbose = true
    }

    /// Log with OODA context
    let private log (phase: OodaPhase) (message: string) (verbose: bool) : unit =
        if verbose then
            let phaseStr =
                match phase with
                | Observe -> "OBSERVE"
                | Orient -> "ORIENT"
                | Decide -> "DECIDE"
                | Act -> "ACT"
                | Verify -> "VERIFY"
            let ts = DateTime.UtcNow.ToString("HH:mm:ss.fff")
            printfn "[%s] [OODA:%-7s] %s" ts phaseStr message

    /// Execute shell command (for health checks)
    let private execCommand (command: string) (args: string) (timeoutMs: int) : (int * string) =
        let psi = Diagnostics.ProcessStartInfo(
            FileName = command,
            Arguments = args,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true
        )
        use proc = new Diagnostics.Process()
        proc.StartInfo <- psi
        proc.Start() |> ignore

        let stdout = proc.StandardOutput.ReadToEnd()
        if proc.WaitForExit(timeoutMs) then
            (proc.ExitCode, stdout)
        else
            proc.Kill()
            (-1, "Timeout")

    /// OBSERVE phase - Collect system state
    let observe (twin: DigitalTwin) (config: SupervisorConfig) : Observation =
        let sw = Stopwatch.StartNew()

        if config.Verbose then
            log Observe "Collecting sensor data..." config.Verbose

        // Collect container health from twin
        let containerHealth =
            twin.Phenotypes
            |> Map.map (fun _ p -> p.Health)

        // Check port availability (5433 for DB, 4317 for OTEL)
        let (dbCode, _) = execCommand "nc" "-z localhost 5433" 1000
        let portsAvailable = dbCode = 0

        // Check database connectivity
        let (pgCode, _) = execCommand "pg_isready" "-h localhost -p 5433" 2000
        let dbConnected = pgCode = 0

        // Collect active connections
        let activeConnections =
            twin.Phenotypes
            |> Map.map (fun _ p -> p.ActiveConnections)

        // Count recent errors
        let recentErrors =
            twin.Phenotypes
            |> Map.toList
            |> List.sumBy (fun (_, p) -> p.Errors.Length)

        // Check SLA compliance (from last boot)
        let slaCompliant =
            match twin.Cache with
            | Some cache ->
                let bootTimes = cache.StartOrder |> List.sumBy (fun w -> w.MaxParallel)
                bootTimes <= 10  // Rough SLA check
            | None -> true

        sw.Stop()

        if config.Verbose then
            log Observe (sprintf "Observation collected in %dms" sw.ElapsedMilliseconds) config.Verbose

        {
            Timestamp = DateTimeOffset.UtcNow
            ContainerHealth = containerHealth
            PortsAvailable = portsAvailable
            DatabaseConnected = dbConnected
            NetworkLatencyMs = 0  // Would measure with ping
            MemoryPressure = 0.0  // Would read from /proc/meminfo
            CpuUtilization = 0.0  // Would read from /proc/stat
            ActiveConnections = activeConnections
            RecentErrors = recentErrors
            SlaCompliant = slaCompliant
        }

    /// ORIENT phase - Analyze situation
    let orient (observation: Observation) (state: SupervisorState) : Orientation =
        let sw = Stopwatch.StartNew()

        if state.Config.Verbose then
            log Orient "Analyzing situation..." state.Config.Verbose

        // Calculate health score
        let totalContainers = observation.ContainerHealth.Count
        let healthyCount =
            observation.ContainerHealth
            |> Map.filter (fun _ h -> h = ContainerHealth.Healthy)
            |> Map.count

        let healthScore =
            if totalContainers = 0 then 0.0
            else float healthyCount / float totalContainers

        // Determine trend from history
        let trend =
            if state.History.Length < 2 then 0
            else
                let lastScore = state.History.[0].Orientation.HealthScore
                if healthScore > lastScore + 0.05 then 1
                elif healthScore < lastScore - 0.05 then -1
                else 0

        // Identify issues
        let issues = ResizeArray<string>()

        if not observation.PortsAvailable then
            issues.Add("Ports not available (5433 or 4317)")

        if not observation.DatabaseConnected then
            issues.Add("Database not connected")

        if observation.RecentErrors > 0 then
            issues.Add(sprintf "%d recent errors detected" observation.RecentErrors)

        if healthScore < state.Config.HealthThreshold then
            issues.Add(sprintf "Health score below threshold (%.2f < %.2f)" healthScore state.Config.HealthThreshold)

        // Identify degraded containers
        let degradedContainers =
            observation.ContainerHealth
            |> Map.filter (fun _ h ->
                match h with
                | ContainerHealth.Healthy -> false
                | _ -> true)
            |> Map.keys
            |> Seq.toList

        // Calculate risk level
        let riskLevel =
            if observation.RecentErrors > 10 || not observation.DatabaseConnected then 4  // Critical
            elif healthScore < 0.5 then 3  // High
            elif healthScore < 0.7 then 2  // Medium
            elif issues.Count > 0 then 1   // Low
            else 0                          // None

        // Predict 5-order effects (simplified)
        let predictedEffects =
            Map.ofList [
                (1, ["Container state changes"])
                (2, ["Health check propagation"])
                (3, ["Dashboard updates"])
                (4, ["SLA metrics impact"])
                (5, ["Compliance reporting"])
            ]

        sw.Stop()

        if state.Config.Verbose then
            log Orient (sprintf "Health=%.2f Trend=%d Risk=%d Issues=%d" healthScore trend riskLevel issues.Count) state.Config.Verbose

        {
            HealthScore = healthScore
            Trend = trend
            Issues = issues |> Seq.toList
            Priorities = if degradedContainers.IsEmpty then [] else ["Restore degraded containers"]
            RiskLevel = riskLevel
            DegradedContainers = degradedContainers
            PredictedEffects = predictedEffects
        }

    /// DECIDE phase - Determine action (with hysteresis per SC-OODA-005)
    let decide (observation: Observation) (orientation: Orientation) (state: SupervisorState) : Decision =
        let sw = Stopwatch.StartNew()

        if state.Config.Verbose then
            log Decide "Evaluating options..." state.Config.Verbose

        // Hysteresis check - don't oscillate decisions
        let shouldApplyHysteresis =
            state.CyclesSinceDecisionChange < state.Config.HysteresisHoldCycles

        // Determine decision
        let decision =
            // Emergency: too many consecutive failures
            if state.ConsecutiveFailures >= state.Config.MaxConsecutiveFailures then
                EmergencyStop "Max consecutive failures exceeded"

            // Critical risk
            elif orientation.RiskLevel >= 4 then
                if not observation.DatabaseConnected then
                    RestartContainer ("indrajaal-db-prod", "Database connection lost")
                else
                    EmergencyStop "Critical risk level detected"

            // High risk with degraded containers
            elif orientation.RiskLevel >= 3 && not orientation.DegradedContainers.IsEmpty then
                let target = orientation.DegradedContainers |> List.head
                RestartContainer (target, "Container degraded")

            // Health below threshold
            elif orientation.HealthScore < state.Config.HealthThreshold - state.Config.HysteresisMargin then
                if shouldApplyHysteresis then
                    match state.LastDecision with
                    | Some (NoAction _) -> NoAction "Hysteresis: waiting for stabilization"
                    | _ -> HealthCheck orientation.DegradedContainers
                else
                    HealthCheck orientation.DegradedContainers

            // Everything healthy
            else
                NoAction "System healthy"

        sw.Stop()

        if state.Config.Verbose then
            let decisionStr =
                match decision with
                | NoAction r -> sprintf "NO_ACTION: %s" r
                | BootMesh _ -> "BOOT_MESH"
                | ShutdownMesh _ -> "SHUTDOWN_MESH"
                | RestartContainer (id, _) -> sprintf "RESTART: %s" id
                | ScaleUp n -> sprintf "SCALE_UP: %d" n
                | ScaleDown n -> sprintf "SCALE_DOWN: %d" n
                | EmergencyStop r -> sprintf "EMERGENCY: %s" r
                | HealthCheck ids -> sprintf "HEALTH_CHECK: %d containers" ids.Length
                | DrainContainer id -> sprintf "DRAIN: %s" id
            log Decide decisionStr state.Config.Verbose

        decision

    /// Guardian validation (SC-CTRL-006)
    let validateWithGuardian (decision: Decision) (config: SupervisorConfig) : Result<Decision, string> =
        if not config.EnableGuardian then
            Ok decision
        else
            // In real implementation: call Guardian.validate_proposal/1
            match decision with
            | EmergencyStop _ ->
                // Emergency always allowed
                Ok decision
            | ShutdownMesh _ ->
                // Require explicit approval for shutdown
                if config.Verbose then
                    log Act "Guardian: Shutdown requires explicit approval" config.Verbose
                Error "Shutdown requires explicit Guardian approval"
            | _ ->
                Ok decision

    /// ACT phase - Execute decision
    let act (decision: Decision) (twin: DigitalTwin) (config: SupervisorConfig) : ActionResult =
        let sw = Stopwatch.StartNew()

        if config.Verbose then
            log Act "Executing decision..." config.Verbose

        // Validate with Guardian
        match validateWithGuardian decision config with
        | Error reason ->
            sw.Stop()
            Vetoed reason

        | Ok validatedDecision ->
            let result =
                match validatedDecision with
                | NoAction reason ->
                    Skipped reason

                | BootMesh bootConfig ->
                    try
                        let result = MeshStartup.boot twin bootConfig
                        if result.AllSucceeded then
                            Success ("Mesh booted successfully", result.TotalDurationMs)
                        else
                            Failure ("Boot failed for some containers", result.TotalDurationMs)
                    with ex ->
                        Failure (ex.Message, sw.ElapsedMilliseconds)

                | ShutdownMesh shutdownConfig ->
                    try
                        let result = MeshShutdown.shutdown twin shutdownConfig
                        if result.AllGraceful then
                            Success ("Mesh shutdown gracefully", result.TotalDurationMs)
                        else
                            Failure ("Shutdown required force kills", result.TotalDurationMs)
                    with ex ->
                        Failure (ex.Message, sw.ElapsedMilliseconds)

                | RestartContainer (id, reason) ->
                    try
                        if config.Verbose then
                            log Act (sprintf "Restarting %s: %s" id reason) config.Verbose
                        // Stop then start
                        let genotype = twin.Genotypes.[id]
                        let (code, _) = execCommand "podman-compose" (sprintf "-f %s restart %s" MeshStartup.defaultConfig.ComposeFile genotype.Name) 30000
                        if code = 0 then
                            Success (sprintf "Container %s restarted" id, sw.ElapsedMilliseconds)
                        else
                            Failure (sprintf "Failed to restart %s" id, sw.ElapsedMilliseconds)
                    with ex ->
                        Failure (ex.Message, sw.ElapsedMilliseconds)

                | EmergencyStop reason ->
                    try
                        if config.Verbose then
                            log Act (sprintf "EMERGENCY STOP: %s" reason) config.Verbose
                        let result = MeshShutdown.emergencyShutdown twin
                        Success (sprintf "Emergency stop: %s" reason, result.TotalDurationMs)
                    with ex ->
                        Failure (ex.Message, sw.ElapsedMilliseconds)

                | HealthCheck ids ->
                    if config.Verbose then
                        log Act (sprintf "Health checking %d containers" ids.Length) config.Verbose
                    Skipped "Health check scheduled"

                | DrainContainer id ->
                    if config.Verbose then
                        log Act (sprintf "Draining %s" id) config.Verbose
                    Skipped "Drain scheduled"

                | ScaleUp n ->
                    try
                        if config.Verbose then
                            log Act (sprintf "Scaling up by %d" n) config.Verbose
                        let currentCount =
                            twin.Phenotypes
                            |> Map.filter (fun k _ -> k.Contains("app"))
                            |> Map.count
                        let targetCount = currentCount + n
                        let composeArgs = sprintf "-f %s up -d --scale indrajaal-ex-app=%d" MeshStartup.defaultConfig.ComposeFile targetCount
                        let (code, output) = execCommand "podman-compose" composeArgs 60000
                        if code = 0 then
                            Success (sprintf "Scaled up to %d app instances" targetCount, sw.ElapsedMilliseconds)
                        else
                            Failure (sprintf "Scale up failed: %s" (output.Trim()), sw.ElapsedMilliseconds)
                    with ex ->
                        Failure (ex.Message, sw.ElapsedMilliseconds)

                | ScaleDown n ->
                    try
                        if config.Verbose then
                            log Act (sprintf "Scaling down by %d" n) config.Verbose
                        let currentCount =
                            twin.Phenotypes
                            |> Map.filter (fun k _ -> k.Contains("app"))
                            |> Map.count
                        let targetCount = max 1 (currentCount - n)
                        if targetCount >= currentCount then
                            Skipped "Already at minimum scale"
                        else
                            let composeArgs = sprintf "-f %s up -d --scale indrajaal-ex-app=%d" MeshStartup.defaultConfig.ComposeFile targetCount
                            let (code, output) = execCommand "podman-compose" composeArgs 60000
                            if code = 0 then
                                Success (sprintf "Scaled down to %d app instances" targetCount, sw.ElapsedMilliseconds)
                            else
                                Failure (sprintf "Scale down failed: %s" (output.Trim()), sw.ElapsedMilliseconds)
                    with ex ->
                        Failure (ex.Message, sw.ElapsedMilliseconds)

            sw.Stop()

            if config.Verbose then
                let resultStr =
                    match result with
                    | Success (m, ms) -> sprintf "SUCCESS: %s (%dms)" m ms
                    | Failure (e, ms) -> sprintf "FAILURE: %s (%dms)" e ms
                    | Skipped r -> sprintf "SKIPPED: %s" r
                    | Vetoed r -> sprintf "VETOED: %s" r
                log Act resultStr config.Verbose

            result

    /// Run single OODA cycle
    let runCycle (state: SupervisorState) : OodaCycle * SupervisorState =
        let cycleStart = DateTimeOffset.UtcNow
        let sw = Stopwatch.StartNew()

        // OBSERVE
        let observation = observe state.Twin state.Config

        // ORIENT
        let orientation = orient observation state

        // DECIDE
        let decision = decide observation orientation state

        // ACT
        let actionResult = act decision state.Twin state.Config

        sw.Stop()

        // Build cycle record
        let cycle = {
            CycleNumber = state.History.Length + 1
            StartTime = cycleStart
            EndTime = DateTimeOffset.UtcNow
            DurationMs = sw.ElapsedMilliseconds
            Observation = observation
            Orientation = orientation
            Decision = decision
            ActionResult = actionResult
            SlaCycleCompliant = sw.ElapsedMilliseconds <= int64 state.Config.MaxCycleDurationMs
        }

        // Update state
        let consecutiveFailures =
            match actionResult with
            | Failure _ -> state.ConsecutiveFailures + 1
            | _ -> 0

        let decisionChanged =
            match state.LastDecision, decision with
            | Some (NoAction _), NoAction _ -> false
            | Some (HealthCheck _), HealthCheck _ -> false
            | Some prev, curr when prev = curr -> false
            | _ -> true

        let newState = {
            state with
                History = cycle :: (state.History |> List.truncate 99)  // Keep last 100
                ConsecutiveFailures = consecutiveFailures
                LastDecision = Some decision
                CyclesSinceDecisionChange =
                    if decisionChanged then 0
                    else state.CyclesSinceDecisionChange + 1
        }

        // Log cycle completion
        if state.Config.Verbose then
            let slaStr = if cycle.SlaCycleCompliant then "✓" else "✗"
            log Verify (sprintf "Cycle #%d complete in %dms [SLA %s]" cycle.CycleNumber cycle.DurationMs slaStr) state.Config.Verbose

        (cycle, newState)

    /// Create initial supervisor state
    let createState (twin: DigitalTwin) (config: SupervisorConfig) : SupervisorState = {
        Twin = twin
        Config = config
        History = []
        CurrentPhase = Observe
        ConsecutiveFailures = 0
        LastDecision = None
        CyclesSinceDecisionChange = 0
        Running = false
    }

    /// Run supervisor loop
    let run (twin: DigitalTwin) (config: SupervisorConfig) : unit =
        printfn ""
        printfn "\u001b[35m\u001b[1m>>> INDRAJAAL OODA BIOMORPHIC SUPERVISOR <<<\u001b[0m"
        printfn ""

        let mutable state = { createState twin config with Running = true }

        while state.Running do
            let (cycle, newState) = runCycle state
            state <- newState

            // Print dashboard summary
            printfn ""
            printfn "╔══════════════════════════════════════════════════════════════╗"
            printfn "║  Cycle: #%-4d  │  Health: %.2f  │  Risk: %d  │  SLA: %s     ║"
                cycle.CycleNumber
                cycle.Orientation.HealthScore
                cycle.Orientation.RiskLevel
                (if cycle.SlaCycleCompliant then "✓" else "✗")
            printfn "╚══════════════════════════════════════════════════════════════╝"
            printfn ""

            // Wait for next cycle
            if state.Running then
                Thread.Sleep(config.CycleIntervalMs)

    /// Run with default config
    let runDefault (twin: DigitalTwin) : unit =
        run twin defaultConfig

File: sub-projects/intelitor-v5.2/native/ignition_daemon/src/ooda_supervisor.rs
use crate::errors::IgnitionError;
use crate::podman;
use crate::digital_twin::{Genotype, Phenotype, TwinDrift, compare_twin};
use std::time::{Duration, Instant};
use tokio::time;
use tracing::{debug, error, info, instrument, warn, span, Level};
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct BootConfig {
    pub mode: String,
    pub max_retries: u32,
    pub timeout_ms: u64,
}

impl Default for BootConfig {
    fn default() -> Self {
        Self {
            mode: "prod".into(),
            max_retries: 3,
            timeout_ms: 30000,
        }
    }
}

#[derive(Debug, Clone)]
pub enum Decision {
    NoAction(String),
    BootMesh(BootConfig),
    ShutdownMesh,
    RestartContainer(String, String),
    ScaleUp(u32),
    ScaleDown(u32),
    EmergencyStop(String),
    HealthCheck(Vec<String>),
    DrainContainer(String),
}

#[derive(Debug, Clone)]
pub struct Observation {
    pub active_containers: usize,
    pub mesh_healthy: bool,
    pub failed_nodes: Vec<String>,
    pub phenotypes: Vec<Phenotype>,
    pub timestamp: Instant,
}

#[derive(Debug, Clone)]
pub struct Orientation {
    pub drift_detected: bool,
    pub twin_drifts: Vec<TwinDrift>,
    pub overload: bool,
    pub missing_critical_nodes: bool,
    pub timestamp: Instant,
}

#[derive(Debug, Clone)]
pub struct ActionResult {
    pub success: bool,
    pub message: String,
    pub timestamp: Instant,
}

#[derive(Debug, Clone)]
pub struct SupervisorConfig {
    pub cycle_interval_ms: u64,
    pub observe_budget_ms: u64,
    pub orient_budget_ms: u64,
    pub decide_budget_ms: u64,
    pub act_budget_ms: u64,
    pub expected_genotypes: Vec<Genotype>,
}

impl Default for SupervisorConfig {
    fn default() -> Self {
        Self {
            cycle_interval_ms: 30, // 30ms OODA loop (Biomorphic Target)
            observe_budget_ms: 30,
            orient_budget_ms: 20,
            decide_budget_ms: 20,
            act_budget_ms: 20,
            expected_genotypes: crate::digital_twin::build_sil6_genotypes(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct SupervisorState {
    pub mesh_running: bool,
    pub current_cycle: u64,
    pub last_health_check: Option<Instant>,
    pub consecutive_failures: u32,
}

#[derive(Debug, Clone)]
pub struct OodaCycle {
    pub cycle_number: u64,
    pub start: Instant,
    pub duration_ms: u64,
    pub observation: Observation,
    pub orientation: Orientation,
    pub decision: Decision,
    pub action_result: ActionResult,
    pub sla_compliant: bool,
}

/// The OODA Supervisor for the SIL-6 Mesh.
///
/// Implements the Observe-Orient-Decide-Act loop under strict <100ms timing constraints.
pub struct OodaSupervisor {
    pub config: SupervisorConfig,
    pub state: SupervisorState,
}

impl OodaSupervisor {
    pub fn new(config: SupervisorConfig) -> Self {
        Self {
            config,
            state: SupervisorState {
                mesh_running: false,
                current_cycle: 0,
                last_health_check: None,
                consecutive_failures: 0,
            },
        }
    }

    /// PHASE 1: OBSERVE
    /// Gathers state from the physical mesh via Podman REST API/CLI.
    #[instrument(skip(self))]
    pub async fn observe(&self) -> Result<Observation, IgnitionError> {
        let start = Instant::now();
        
        // Gather stats
        let stats = podman::get_all_stats().await?;
        let active_containers = stats.len();
        
        let mut failed_nodes = Vec::new();
        let mut phenotypes = Vec::new();

        for stat in &stats {
            // Inspect container for actual image (best-effort, fallback to "unknown")
            let actual_image = podman::podman_inspect(&stat.name, "{{.Config.Image}}")
                .await
                .unwrap_or_else(|_| "unknown".into());

            let running = podman::podman_inspect(&stat.name, "{{.State.Running}}")
                .await
                .map(|s| s.trim() == "true")
                .unwrap_or(false);

            if !running {
                failed_nodes.push(stat.name.clone());
            }

            phenotypes.push(Phenotype {
                container_name: stat.name.clone(),
                actual_image,
                actual_ports: vec![],
                actual_env: HashMap::new(),
                running,
            });
        }
        
        let elapsed = start.elapsed().as_millis() as u64;
        if elapsed > self.config.observe_budget_ms {
            warn!("Observe phase exceeded budget: {}ms > {}ms", elapsed, self.config.observe_budget_ms);
        }
        
        Ok(Observation {
            active_containers,
            mesh_healthy: active_containers > 0 && failed_nodes.is_empty(),
            failed_nodes,
            phenotypes,
            timestamp: Instant::now(),
        })
    }

    /// PHASE 2: ORIENT
    /// Compares observation against the Digital Twin Genotypes to find drifts.
    #[instrument(skip(self))]
    pub fn orient(&self, obs: &Observation) -> Orientation {
        let start = Instant::now();
        
        let mut twin_drifts = Vec::new();
        let mut missing_critical_nodes = false;

        for expected in &self.config.expected_genotypes {
            if let Some(actual) = obs.phenotypes.iter().find(|p| p.container_name == expected.container_name) {
                let drift = compare_twin(expected, actual);
                if !drift.is_aligned {
                    twin_drifts.push(drift);
                }
            } else {
                missing_critical_nodes = true;
            }
        }
        
        let elapsed = start.elapsed().as_millis() as u64;
        if elapsed > self.config.orient_budget_ms {
            warn!("Orient phase exceeded budget: {}ms > {}ms", elapsed, self.config.orient_budget_ms);
        }
        
        Orientation {
            drift_detected: !twin_drifts.is_empty(),
            twin_drifts,
            overload: false,
            missing_critical_nodes,
            timestamp: Instant::now(),
        }
    }

    /// PHASE 3: DECIDE
    /// Forms a decision based on the current orientation.
    #[instrument(skip(self))]
    pub fn decide(&self, obs: &Observation, orient: &Orientation) -> Decision {
        let start = Instant::now();

        let decision = if let Some(rule_decision) = crate::rule_engine::evaluate_decision(obs, orient, &self.state) {
            info!("Rule Engine Decision: {:?}", rule_decision);
            rule_decision
        } else if !self.state.mesh_running && orient.missing_critical_nodes {
            Decision::BootMesh(BootConfig::default())
        } else if self.state.mesh_running && orient.missing_critical_nodes {
            Decision::EmergencyStop("Missing critical nodes while running".into())
        } else if orient.drift_detected {
            // For now, take the first drifted container and restart it
            if let Some(drift) = orient.twin_drifts.first() {
                Decision::RestartContainer(drift.container_name.clone(), "Drift detected".into())
            } else {
                Decision::NoAction("Drift detected but no target identified".into())
            }
        } else {
            Decision::NoAction("Mesh is fully aligned with genotype".into())
        };
        
        let elapsed = start.elapsed().as_millis() as u64;
        if elapsed > self.config.decide_budget_ms {
            warn!("Decide phase exceeded budget: {}ms > {}ms", elapsed, self.config.decide_budget_ms);
        }
        
        decision
    }

    /// PHASE 4: ACT
    /// Executes the decision using the Podman commands.
    #[instrument(skip(self))]
    pub async fn act(&mut self, decision: Decision) -> Result<ActionResult, IgnitionError> {
        let start = Instant::now();
        
        let mut result = ActionResult {
            success: true,
            message: "Action completed".into(),
            timestamp: Instant::now(),
        };

        match decision {
            Decision::NoAction(reason) => {
                result.message = format!("No action needed: {}", reason);
            }
            Decision::BootMesh(config) => {
                self.state.mesh_running = true;
                result.message = format!("Booting mesh in {} mode", config.mode);
                // Implementation would call robust_launch::launch_mesh()
            }
            Decision::ShutdownMesh => {
                self.state.mesh_running = false;
                result.message = "Mesh shutting down".into();
            }
            Decision::RestartContainer(container, reason) => {
                info!("Restarting container {} due to: {}", container, reason);
                podman::stop_container(&container, 10).await?;
                podman::start_container(&container).await?;
                result.message = format!("Restarted {}", container);
            }
            Decision::ScaleUp(count) => {
                result.message = format!("Scaled up by {}", count);
            }
            Decision::ScaleDown(count) => {
                result.message = format!("Scaled down by {}", count);
            }
            Decision::EmergencyStop(reason) => {
                self.state.mesh_running = false;
                result.message = format!("Emergency stop initiated: {}", reason);
            }
            Decision::HealthCheck(nodes) => {
                result.message = format!("Triggered health checks for {} nodes", nodes.len());
            }
            Decision::DrainContainer(container) => {
                // LLM escalation: ask OpenRouter for advice before acting
                let prompt = format!(
                    "Container '{}' has drifted from genotype. Cycle #{}, mesh_running={}. \
                     Should we: (a) drain and restart, (b) restart only, (c) ignore? \
                     Respond with one letter and a brief reason.",
                    container, self.state.current_cycle, self.state.mesh_running
                );
                match crate::openrouter::query_llm_advisor(&prompt).await {
                    Ok(advice) => {
                        info!("LLM Advisor for {}: {}", container, advice);
                        result.message = format!("LLM advised on {}: {}", container, advice);
                    }
                    Err(e) => {
                        warn!("LLM unavailable ({}), proceeding with drain for {}", e, container);
                    }
                }
                // Execute drain + restart
                if let Err(e) = podman::stop_container(&container, 10).await {
                    warn!("Failed to drain {}: {}", container, e);
                    result.success = false;
                }
                if let Err(e) = podman::start_container(&container).await {
                    warn!("Failed to restart {}: {}", container, e);
                    result.success = false;
                }
                if result.success {
                    result.message = format!("Drained and restarted {} (LLM-guided)", container);
                }
            }
        }
        
        let elapsed = start.elapsed().as_millis() as u64;
        if elapsed > self.config.act_budget_ms {
            warn!("Act phase exceeded budget: {}ms > {}ms", elapsed, self.config.act_budget_ms);
        }
        
        result.timestamp = Instant::now();
        Ok(result)
    }

    /// VALIDATE
    /// Guardian validation step before committing to action.
    /// P0 decisions (EmergencyStop, BootMesh) require Guardian approval.
    /// Fail-closed in release builds; permissive in debug/test (SC-GUARD-002).
    pub fn validate_with_guardian(&self, decision: &Decision) -> bool {
        match decision {
            Decision::EmergencyStop(_) | Decision::BootMesh(_) => {
                warn!("Guardian gate: P0 decision {:?} requires approval", decision);
                // In debug/test: allow. In release: fail-closed (require Guardian service).
                cfg!(debug_assertions)
            }
            _ => true,
        }
    }

    /// Run a complete OODA cycle
    pub async fn run_cycle(&mut self) -> Result<OodaCycle, IgnitionError> {
        let cycle_start = Instant::now();
        self.state.current_cycle += 1;

        let obs = self.observe().await?;
        let orient = self.orient(&obs);
        let decision = self.decide(&obs, &orient);

        let final_decision = if self.validate_with_guardian(&decision) {
            decision
        } else {
            warn!("Guardian rejected decision: {:?}", decision);
            Decision::NoAction("Guardian rejected decision".into())
        };

        let act_res = self.act(final_decision.clone()).await?;

        let duration_ms = cycle_start.elapsed().as_millis() as u64;
        let sla_compliant = duration_ms <= 100;

        if !sla_compliant {
            warn!("OODA cycle {} violated SLA. Duration: {}ms", self.state.current_cycle, duration_ms);
        }

        Ok(OodaCycle {
            cycle_number: self.state.current_cycle,
            start: cycle_start,
            duration_ms,
            observation: obs,
            orientation: orient,
            decision: final_decision,
            action_result: act_res,
            sla_compliant,
        })
    }

    /// Run a shadow OODA cycle (no real actions)
    pub async fn run_shadow_cycle(&mut self) -> Result<OodaCycle, IgnitionError> {
        let cycle_start = Instant::now();
        self.state.current_cycle += 1;

        let obs = self.observe().await?;
        let orient = self.orient(&obs);
        let decision = self.decide(&obs, &orient);

        let final_decision = if self.validate_with_guardian(&decision) {
            decision
        } else {
            Decision::NoAction("Guardian rejected decision".into())
        };

        let act_res = ActionResult {
            success: true,
            message: format!("Shadow mode: Would have executed {:?}", final_decision),
            timestamp: Instant::now(),
        };

        let duration_ms = cycle_start.elapsed().as_millis() as u64;
        let sla_compliant = duration_ms <= 100;

        if !sla_compliant {
            warn!("Shadow OODA cycle {} violated SLA. Duration: {}ms", self.state.current_cycle, duration_ms);
        }

        Ok(OodaCycle {
            cycle_number: self.state.current_cycle,
            start: cycle_start,
            duration_ms,
            observation: obs,
            orientation: orient,
            decision: final_decision,
            action_result: act_res,
            sla_compliant,
        })
    }

    /// Start a continuous OODA loop
    pub async fn start_loop(&mut self) -> Result<(), IgnitionError> {
        let mut interval = time::interval(Duration::from_millis(self.config.cycle_interval_ms));
        
        loop {
            interval.tick().await;
            
            match self.run_cycle().await {
                Ok(cycle) => {
                    debug!("Cycle {} completed in {}ms", cycle.cycle_number, cycle.duration_ms);
                }
                Err(e) => {
                    error!("Cycle failed: {}", e);
                    self.state.consecutive_failures += 1;
                    if self.state.consecutive_failures > 5 {
                        return Err(IgnitionError::InternalError("Too many OODA cycle failures".into()));
                    }
                }
            }
        }
    }
}
